File: code/backend/java/core/src/main/java/games/paths/core/service/story/StoryQueryService.java
package games.paths.core.service.story;

import games.paths.core.entity.story.CardEntity;
import games.paths.core.entity.story.CharacterTemplateEntity;
import games.paths.core.entity.story.ClassBonusEntity;
import games.paths.core.entity.story.ClassEntity;
import games.paths.core.entity.story.StoryDifficultyEntity;
import games.paths.core.entity.story.StoryEntity;
import games.paths.core.entity.story.TextEntity;
import games.paths.core.entity.story.TraitEntity;
import games.paths.core.model.story.CardInfo;
import games.paths.core.model.story.CharacterTemplateInfo;
import games.paths.core.model.story.ClassBonusInfo;
import games.paths.core.model.story.ClassInfo;
import games.paths.core.model.story.DifficultyInfo;
import games.paths.core.model.story.StoryDetail;
import games.paths.core.model.story.StorySummary;
import games.paths.core.model.story.TraitInfo;
import games.paths.core.port.story.StoryQueryPort;
import games.paths.core.port.story.StoryReadPort;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * StoryQueryService - Domain service implementing story querying.
 * Reads story data from the persistence layer and maps it to domain models.
 * Ports are injected via constructor by the launcher configuration.
 *
 * <p>Enhanced in Step 15 with category/group listing, filtering,
 * and enriched story detail with character templates, classes, traits, and card info.</p>
 */
public class StoryQueryService implements StoryQueryPort {

    private final StoryReadPort readPort;

    public StoryQueryService(StoryReadPort readPort) {
        this.readPort = readPort;
    }

    @Override
    public List<StorySummary> listPublicStories(String lang) {
        List<StoryEntity> stories = readPort.findStoriesByVisibility("PUBLIC");
        return mapToSummaries(stories, lang);
    }

    @Override
    public List<StorySummary> listAllStories(String lang) {
        List<StoryEntity> stories = readPort.findAllStories();
        return mapToSummaries(stories, lang);
    }

    @Override
    public StoryDetail getStoryByUuid(String uuid, String lang) {
        if (uuid == null || uuid.isBlank()) {
            return null;
        }

        Optional<StoryEntity> opt = readPort.findStoryByUuid(uuid);
        if (opt.isEmpty()) {
            return null;
        }

        StoryEntity story = opt.get();
        String title = resolveText(story.getId(), story.getIdTextTitle(), lang);
        String description = resolveText(story.getId(), story.getIdTextDescription(), lang);
        String copyrightText = resolveText(story.getId(), story.getIdTextCopyright(), lang);
        String clockSingular = resolveText(story.getId(), story.getIdTextClockSingular(), lang);
        String clockPlural = resolveText(story.getId(), story.getIdTextClockPlural(), lang);

        List<StoryDifficultyEntity> diffEntities = readPort.findDifficultiesByStoryId(story.getId());
        List<DifficultyInfo> difficulties = new ArrayList<>();
        for (StoryDifficultyEntity diff : diffEntities) {
            String diffDesc = resolveText(story.getId(), diff.getIdTextDescription(), lang);
            difficulties.add(DifficultyInfo.builder()
                    .uuid(diff.getUuid())
                    .description(diffDesc)
                    .expCost(diff.getExpCost() != null ? diff.getExpCost() : 5)
                    .maxWeight(diff.getMaxWeight() != null ? diff.getMaxWeight() : 10)
                    .minCharacter(diff.getMinCharacter() != null ? diff.getMinCharacter() : 1)
                    .maxCharacter(diff.getMaxCharacter() != null ? diff.getMaxCharacter() : 4)
                    .costHelpComa(diff.getCostHelpComa() != null ? diff.getCostHelpComa() : 3)
                    .costMaxCharacteristics(diff.getCostMaxCharacteristics() != null ? diff.getCostMaxCharacteristics() : 3)
                    .numberMaxFreeAction(diff.getNumberMaxFreeAction() != null ? diff.getNumberMaxFreeAction() : 1)
                    .life(diff.getLife() != null ? diff.getLife() : 100)
                    .energy(diff.getEnergy() != null ? diff.getEnergy() : 100)
                    .sad(diff.getSad() != null ? diff.getSad() : 0)
                    .dexterity(diff.getDexterity() != null ? diff.getDexterity() : 10)
                    .intelligence(diff.getIntelligence() != null ? diff.getIntelligence() : 10)
                    .constitution(diff.getConstitution() != null ? diff.getConstitution() : 10)
                    .weight(diff.getWeight() != null ? diff.getWeight() : 10)
                    .idCard(diff.getIdCard())
                    .card(resolveCardInfo(story.getId(), diff.getIdCard(), lang))
                    .traitCostPositiveBudget(diff.getTraitCostPositiveBudget())
                    .traitCostNegativeBudget(diff.getTraitCostNegativeBudget())
                    .build());
        }

        long locationCount = readPort.countLocationsByStoryId(story.getId());
        long eventCount = readPort.countEventsByStoryId(story.getId());
        long itemCount = readPort.countItemsByStoryId(story.getId());

        // Step 15: Character templates, classes, and traits
        List<CharacterTemplateEntity> ctEntities = readPort.findCharacterTemplatesByStoryId(story.getId());
        List<CharacterTemplateInfo> characterTemplates = new ArrayList<>();
        for (CharacterTemplateEntity ct : ctEntities) {
            String ctName = resolveText(story.getId(), ct.getIdTextName(), lang);
            String ctDesc = resolveText(story.getId(), ct.getIdTextDescription(), lang);
            characterTemplates.add(new CharacterTemplateInfo(
                    ct.getUuid(),
                    ctName,
                    ctDesc,
                    ct.getLifeMax() != null ? ct.getLifeMax() : 10,
                    ct.getEnergyMax() != null ? ct.getEnergyMax() : 10,
                    ct.getSadMax() != null ? ct.getSadMax() : 10,
                    ct.getDexterityStart() != null ? ct.getDexterityStart() : 1,
                    ct.getIntelligenceStart() != null ? ct.getIntelligenceStart() : 1,
                    ct.getConstitutionStart() != null ? ct.getConstitutionStart() : 1,
                    ct.getIdCard(),
                    resolveCardInfo(story.getId(), ct.getIdCard(), lang),
                    ct.getIdClassPermitted(),
                    ct.getIdClassProhibited()));
        }

        List<ClassEntity> classEntities = readPort.findClassesByStoryId(story.getId());
        List<ClassBonusEntity> classBonusEntities = readPort.findClassBonusesByStoryId(story.getId());
        List<ClassInfo> classes = new ArrayList<>();
        for (ClassEntity cl : classEntities) {
            String clName = resolveText(story.getId(), cl.getIdTextName(), lang);
            String clDesc = resolveText(story.getId(), cl.getIdTextDescription(), lang);
            List<ClassBonusInfo> bonusList = new ArrayList<>();
            for (ClassBonusEntity cb : classBonusEntities) {
                if (cb.getIdClass() != null && cl.getId() != null
                        && cb.getIdClass().longValue() == cl.getId().longValue()) {
                    bonusList.add(new ClassBonusInfo(
                            cb.getUuid(),
                            cb.getStatistic(),
                            cb.getValue() != null ? cb.getValue() : 0));
                }
            }
            classes.add(new ClassInfo(
                    cl.getId(),
                    cl.getUuid(),
                    clName,
                    clDesc,
                    cl.getWeightMax() != null ? cl.getWeightMax() : 10,
                    cl.getDexterityBase() != null ? cl.getDexterityBase() : 1,
                    cl.getIntelligenceBase() != null ? cl.getIntelligenceBase() : 1,
                    cl.getConstitutionBase() != null ? cl.getConstitutionBase() : 1,
                    cl.getIdCard(),
                    resolveCardInfo(story.getId(), cl.getIdCard(), lang),
                    bonusList));
        }

        List<TraitEntity> traitEntities = readPort.findTraitsByStoryId(story.getId());
        List<TraitInfo> traits = new ArrayList<>();
        for (TraitEntity tr : traitEntities) {
            traits.add(toTraitInfo(story.getId(), tr, lang));
        }

        // Step 15: Resolve card info
        CardInfo cardInfo = resolveCardInfo(story.getId(), story.getIdCard(), lang);

        return StoryDetail.builder()
                .uuid(story.getUuid())
                .title(title)
                .description(description)
                .author(story.getAuthor())
                .category(story.getCategory())
                .group(story.getGroup())
                .visibility(story.getVisibility())
                .priority(story.getPriority() != null ? story.getPriority() : 0)
                .peghi(story.getPeghi() != null ? story.getPeghi() : 0)
                .versionMin(story.getVersionMin())
                .versionMax(story.getVersionMax())
                .clockSingularDescription(clockSingular)
                .clockPluralDescription(clockPlural)
                .idTextClockSingular(story.getIdTextClockSingular())
                .idTextClockPlural(story.getIdTextClockPlural())
                .copyrightText(copyrightText)
                .linkCopyright(story.getLinkCopyright())
                .locationCount((int) locationCount)
                .eventCount((int) eventCount)
                .itemCount((int) itemCount)
                .classCount(classes.size())
                .characterTemplateCount(characterTemplates.size())
                .traitCount(traits.size())
                .difficulties(difficulties)
                .characterTemplates(characterTemplates)
                .classes(classes)
                .traits(traits)
                .card(cardInfo)
                .build();
    }

    // === Step 15: Category and Group queries ===

    @Override
    public List<String> listCategories() {
        return readPort.findDistinctCategoriesByVisibility("PUBLIC");
    }

    @Override
    public List<StorySummary> listStoriesByCategory(String category, String lang) {
        if (category == null || category.isBlank()) {
            return List.of();
        }
        List<StoryEntity> stories = readPort.findStoriesByCategoryAndVisibility(category, "PUBLIC");
        return mapToSummaries(stories, lang);
    }

    @Override
    public List<String> listGroups() {
        return readPort.findDistinctGroupsByVisibility("PUBLIC");
    }

    @Override
    public List<StorySummary> listStoriesByGroup(String group, String lang) {
        if (group == null || group.isBlank()) {
            return List.of();
        }
        List<StoryEntity> stories = readPort.findStoriesByGroupAndVisibility(group, "PUBLIC");
        return mapToSummaries(stories, lang);
    }

    // === Step 23: Trait listing filtered by class ===

    @Override
    public TraitsForClassResult listTraitsForClass(String storyUuid, String classUuid, String lang) {
        if (storyUuid == null || storyUuid.isBlank()) {
            return TraitsForClassResult.storyNotFound();
        }
        Optional<StoryEntity> storyOpt = readPort.findStoryByUuid(storyUuid);
        if (storyOpt.isEmpty()) {
            return TraitsForClassResult.storyNotFound();
        }
        StoryEntity story = storyOpt.get();

        if (classUuid == null || classUuid.isBlank()) {
            return TraitsForClassResult.classNotFound();
        }
        Optional<ClassEntity> classOpt = readPort.findClassByStoryIdAndUuid(story.getId(), classUuid);
        if (classOpt.isEmpty()) {
            return TraitsForClassResult.classNotFound();
        }
        Long classId = classOpt.get().getId();

        List<TraitInfo> traits = new ArrayList<>();
        for (TraitEntity tr : readPort.findTraitsByStoryId(story.getId())) {
            if (isTraitSelectableForClass(tr, classId)) {
                traits.add(toTraitInfo(story.getId(), tr, lang));
            }
        }
        return TraitsForClassResult.ok(traits);
    }

    private static boolean isTraitSelectableForClass(TraitEntity tr, Long classId) {
        Integer permitted = tr.getIdClassPermitted();
        Integer prohibited = tr.getIdClassProhibited();
        boolean permittedOk = permitted == null
                || (classId != null && permitted.longValue() == classId);
        boolean prohibitedOk = prohibited == null
                || classId == null || prohibited.longValue() != classId;
        return permittedOk && prohibitedOk;
    }

    // === Private helpers ===

    private TraitInfo toTraitInfo(Long storyId, TraitEntity tr, String lang) {
        String trName = resolveText(storyId, tr.getIdTextName(), lang);
        String trDesc = resolveText(storyId, tr.getIdTextDescription(), lang);
        return TraitInfo.builder()
                .uuid(tr.getUuid())
                .name(trName)
                .description(trDesc)
                .costPositive(tr.getCostPositive() != null ? tr.getCostPositive() : 0)
                .costNegative(tr.getCostNegative() != null ? tr.getCostNegative() : 0)
                .idClassPermitted(tr.getIdClassPermitted())
                .idClassProhibited(tr.getIdClassProhibited())
                .idCard(tr.getIdCard())
                .card(resolveCardInfo(storyId, tr.getIdCard(), lang))
                // v0.35.2 — reported on BOTH projections; nothing is filtered out here.
                .hideOnStartMatch(tr.isHiddenOnStartMatch())
                .life(tr.getLife() != null ? tr.getLife() : 0)
                .energy(tr.getEnergy() != null ? tr.getEnergy() : 0)
                .sad(tr.getSad() != null ? tr.getSad() : 0)
                .dexterity(tr.getDexterity() != null ? tr.getDexterity() : 0)
                .intelligence(tr.getIntelligence() != null ? tr.getIntelligence() : 0)
                .constitution(tr.getConstitution() != null ? tr.getConstitution() : 0)
                .weight(tr.getWeight() != null ? tr.getWeight() : 0)
                .build();
    }

    private List<StorySummary> mapToSummaries(List<StoryEntity> stories, String lang) {
        List<StorySummary> summaries = new ArrayList<>();
        for (StoryEntity story : stories) {
            String title = resolveText(story.getId(), story.getIdTextTitle(), lang);
            String description = resolveText(story.getId(), story.getIdTextDescription(), lang);

            List<StoryDifficultyEntity> diffs = readPort.findDifficultiesByStoryId(story.getId());
            CardInfo cardInfo = resolveCardInfo(story.getId(), story.getIdCard(), lang);

            summaries.add(new StorySummary(
                    story.getUuid(),
                    title,
                    description,
                    story.getAuthor(),
                    story.getCategory(),
                    story.getGroup(),
                    story.getVisibility(),
                    story.getPriority() != null ? story.getPriority() : 0,
                    story.getPeghi() != null ? story.getPeghi() : 0,
                    diffs.size(),
                    cardInfo));
        }
        return summaries;
    }

    /**
     * Resolves a text entry by story ID, text ID and language.
     * Falls back to "en" if the requested language is not found,
     * and returns null if no text exists at all.
     */
    private String resolveText(Long storyId, Integer idText, String lang) {
        if (idText == null) {
            return null;
        }
        String effectiveLang = (lang != null && !lang.isBlank()) ? lang : "en";

        Optional<TextEntity> text = readPort.findTextByStoryIdTextAndLang(storyId, idText, effectiveLang);
        if (text.isPresent()) {
            return text.get().getShortText();
        }

        // Fallback to English if requested lang not found
        if (!"en".equals(effectiveLang)) {
            Optional<TextEntity> fallback = readPort.findTextByStoryIdTextAndLang(storyId, idText, "en");
            if (fallback.isPresent()) {
                return fallback.get().getShortText();
            }
        }
        return null;
    }

    /**
     * Resolves card info from card entity, including text resolution for card title.
     */
    private CardInfo resolveCardInfo(Long storyId, Integer idCard, String lang) {
        if (idCard == null) {
            return null;
        }
        Optional<CardEntity> cardOpt = readPort.findCardByStoryIdAndCardId(storyId, idCard.longValue());
        if (cardOpt.isEmpty()) {
            return null;
        }
        CardEntity card = cardOpt.get();
        String cardTitle = resolveText(storyId, card.getIdTextTitle(), lang);
        String cardDescription = resolveText(storyId, card.getIdTextDescription(), lang);
        String cardCopyrightText = resolveText(storyId, card.getIdTextCopyright(), lang);
        return new CardInfo(
                card.getUuid(),
                card.getCardType(),
                card.getUrlImage(),
                card.getAlternativeImage(),
                card.getAwesomeIcon(),
                card.getStyleMain(),
                card.getStyleDetail(),
                card.getStyleImageLittle(),
                card.getStyleImageMedium(),
                card.getStyleImageLarge(),
                cardTitle,
                cardDescription,
                cardCopyrightText,
                card.getLinkCopyright(),
                null);
    }
}

File: code/backend/node/src/core/services/StoryQueryService.ts
import type { PrismaClient } from '@prisma/client';
import type { StoryRepository } from '../ports/StoryRepository';
import type { DifficultyRepository } from '../ports/DifficultyRepository';
import type { StorySummaryResponse } from '../models/Story';
import { resolveText, toInt } from '../../adapters/persistence/prisma/textResolver';

export class StoryQueryService {
  constructor(
    private storyRepo: StoryRepository,
    private difficultyRepo: DifficultyRepository,
    private prisma: PrismaClient,
  ) {}

  /** Build the public card object (or null) for a list_cards row, resolving texts. */
  private async _cardResponse(card: any, lang: string): Promise<any | null> {
    if (!card) return null;
    return {
      uuid: card.uuid,
      urlImage: card.urlImmage ?? null,
      alternativeImage: card.alternativeImage ?? null,
      awesomeIcon: card.awesomeIcon ?? null,
      styleMain: card.styleMain ?? null,
      styleDetail: card.styleDetail ?? null,
      title: (await resolveText(this.prisma, card.idStory, card.idTextTitle, lang)) ?? null,
      description: (await resolveText(this.prisma, card.idStory, card.idTextDescription, lang)) ?? null,
      linkCopyright: card.linkCopyright ?? null,
    };
  }

  /** Resolve the story-level primary card (idCard FK, else first card with an image). */
  private async _primaryCard(idStory: number, idCard: number | null | undefined, lang: string): Promise<any | null> {
    let card: any = null;
    if (idCard != null) {
      // story.id_card references list_cards(id) (the card PK), not the id_card column.
      card = await this.prisma.card.findFirst({ where: { idStory, id: idCard } });
    }
    if (!card) {
      card =
        (await this.prisma.card.findFirst({ where: { idStory, NOT: { urlImmage: null } }, orderBy: { id: 'asc' } })) ??
        (await this.prisma.card.findFirst({ where: { idStory }, orderBy: { id: 'asc' } }));
    }
    return this._cardResponse(card, lang);
  }

  /** Resolve a nested entity's card by its idCard FK (nullable). */
  private async _cardByIdCard(idStory: number, idCard: number | null | undefined, lang: string): Promise<any | null> {
    if (idCard == null) return null;
    // id_card FK references list_cards(id) (the card PK).
    const card = await this.prisma.card.findFirst({ where: { idStory, id: idCard } });
    return this._cardResponse(card, lang);
  }

  private async _summaries(stories: any[], lang: string): Promise<StorySummaryResponse[]> {
    return Promise.all(
      stories.map(async (s) => ({
        uuid: s.uuid,
        title: s.title,
        description: s.description || null,
        author: s.author || null,
        category: s.category || null,
        group: s.group || null,
        visibility: s.visibility,
        priority: s.priority || 0,
        peghi: s.peghi || 0,
        difficultyCount: 0,
        card: await this._primaryCard(s.idStory ?? toInt(s.id), s.idCard, lang),
      })),
    );
  }

  async listPublicStories(lang = 'en'): Promise<StorySummaryResponse[]> {
    return this._summaries(await this.storyRepo.listPublic(), lang);
  }

  async getStoryByUuid(uuid: string, lang = 'en') {
    const story = await this.storyRepo.findByUuid(uuid);
    if (!story) return null;
    const idStory = story.idStory ?? toInt(story.id);

    const [difficulties, classes, traits, templates, classBonuses, card] = await Promise.all([
      this.difficultyRepo.findByStoryUuid(uuid),
      this.prisma.storyClass.findMany({ where: { idStory }, orderBy: { id: 'asc' } }),
      this.prisma.trait.findMany({ where: { idStory }, orderBy: { id: 'asc' } }),
      this.prisma.characterTemplate.findMany({ where: { idStory }, orderBy: { idTipo: 'asc' } }),
      this.prisma.classBonus.findMany({ where: { idStory }, orderBy: { id: 'asc' } }),
      this._primaryCard(idStory, story.idCard, lang),
    ]);

    // Group class bonuses by their integer class id (list_classes_bonus.id_class)
    const bonusesByClass = new Map<number, any[]>();
    for (const b of classBonuses) {
      const list = bonusesByClass.get(b.idClass) ?? [];
      list.push({ uuid: b.uuid, statistic: b.statistic, value: b.value });
      bonusesByClass.set(b.idClass, list);
    }

    const classesOut = await Promise.all(
      classes.map(async (c) => ({
        id: c.id,
        idClass: c.id,
        uuid: c.uuid,
        title: (await resolveText(this.prisma, idStory, c.idTextName, lang)) || `Class ${c.id}`,
        name: (await resolveText(this.prisma, idStory, c.idTextName, lang)) || `Class ${c.id}`,
        description: (await resolveText(this.prisma, idStory, c.idTextDescription, lang)) ?? null,
        idCard: c.idCard ?? null,
        card: await this._cardByIdCard(idStory, c.idCard, lang),
        weightMax: c.weightMax,
        dexterityBase: c.dexterityBase,
        intelligenceBase: c.intelligenceBase,
        constitutionBase: c.constitutionBase,
        bonuses: bonusesByClass.get(c.id) ?? [],
      })),
    );

    const traitsOut = await Promise.all(traits.map((t) => this._traitOut(idStory, t, lang)));

    const templatesOut = await Promise.all(
      templates.map(async (tpl) => ({
        uuid: tpl.uuid,
        title: (await resolveText(this.prisma, idStory, tpl.idTextName, lang)) || `Template ${tpl.idTipo}`,
        name: (await resolveText(this.prisma, idStory, tpl.idTextName, lang)) || `Template ${tpl.idTipo}`,
        description: (await resolveText(this.prisma, idStory, tpl.idTextDescription, lang)) ?? null,
        idCard: tpl.idCard ?? null,
        card: await this._cardByIdCard(idStory, tpl.idCard, lang),
        lifeMax: tpl.lifeMax,
        energyMax: tpl.energyMax,
        sadMax: tpl.sadMax,
        dexterityStart: tpl.dexterityStart,
        intelligenceStart: tpl.intelligenceStart,
        constitutionStart: tpl.constitutionStart,
        idClassPermitted: tpl.idClassPermitted ?? null,
        idClassProhibited: tpl.idClassProhibited ?? null,
      })),
    );

    const difficultiesOut = await Promise.all(
      difficulties.map(async (d: any) => ({
        uuid: d.uuid,
        title: d.title,
        name: d.title,
        description: d.description || null,
        level: d.level ?? 1,
        idCard: d.idCard ?? null,
        card: await this._cardByIdCard(idStory, d.idCard, lang),
        life: d.life,
        energy: d.energy,
        sad: d.sad,
        dexterity: d.dexterity,
        intelligence: d.intelligence,
        constitution: d.constitution,
        weight: d.weight,
        traitCostPositiveBudget: d.traitCostPositiveBudget ?? null,
        traitCostNegativeBudget: d.traitCostNegativeBudget ?? null,
      })),
    );

    return {
      uuid: story.uuid,
      title: story.title,
      description: story.description || null,
      author: story.author || null,
      category: story.category || null,
      group: story.group || null,
      visibility: story.visibility,
      priority: story.priority || 0,
      peghi: story.peghi || 0,
      difficultyCount: difficulties.length,
      card,
      difficulties: difficultiesOut,
      locationCount: 0,
      eventCount: 0,
      itemCount: 0,
      classCount: classes.length,
      characterTemplateCount: templates.length,
      traitCount: traits.length,
      classes: classesOut,
      traits: traitsOut,
      characterTemplates: templatesOut,
    };
  }

  private async _traitOut(idStory: number, t: any, lang: string) {
    const name = (await resolveText(this.prisma, idStory, t.idTextName, lang)) || `Trait ${t.id}`;
    return {
      uuid: t.uuid,
      title: name,
      name,
      description: (await resolveText(this.prisma, idStory, t.idTextDescription, lang)) ?? null,
      idCard: t.idCard ?? null,
      card: await this._cardByIdCard(idStory, t.idCard, lang),
      costPositive: t.costPositive,
      costNegative: t.costNegative,
      life: t.life,
      energy: t.energy,
      sad: t.sad,
      dexterity: t.dexterity,
      intelligence: t.intelligence,
      constitution: t.constitution,
      weight: t.weight,
      idClassPermitted: t.idClassPermitted ?? null,
      idClassProhibited: t.idClassProhibited ?? null,
    };
  }

  async listCategories(): Promise<string[]> {
    return this.storyRepo.listCategories();
  }

  async listStoriesByCategory(category: string, lang = 'en'): Promise<StorySummaryResponse[]> {
    return this._summaries(await this.storyRepo.listByCategory(category), lang);
  }

  async listGroups(): Promise<string[]> {
    return this.storyRepo.listGroups();
  }

  async listStoriesByGroup(group: string, lang = 'en'): Promise<StorySummaryResponse[]> {
    return this._summaries(await this.storyRepo.listByGroup(group), lang);
  }

  /**
   * Get traits for a specific class within a story. Class compatibility uses the
   * scoped integer class id (list_traits.id_class_permitted / prohibited).
   */
  async getTraitsForClass(storyUuid: string, classUuid: string) {
    const story = await this.storyRepo.findByUuid(storyUuid);
    if (!story) return { error: 'STORY_NOT_FOUND' };
    const idStory = story.idStory ?? toInt(story.id);

    const cls = await this.prisma.storyClass.findUnique({ where: { uuid: classUuid } });
    if (!cls || cls.idStory !== idStory) return { error: 'CLASS_NOT_FOUND' };

    const classId = cls.id;
    const allTraits = await this.prisma.trait.findMany({ where: { idStory }, orderBy: { id: 'asc' } });
    const compatible = allTraits.filter((t) => {
      const permittedOk = t.idClassPermitted == null || t.idClassPermitted === classId;
      const prohibitedOk = t.idClassProhibited == null || t.idClassProhibited !== classId;
      return permittedOk && prohibitedOk;
    });

    return { traits: await Promise.all(compatible.map((t) => this._traitOut(idStory, t, 'en'))) };
  }
}
